### PR TITLE
[MNT] EEGNetv4 FIX

### PR DESCRIPTION
**Pull request by bruAristimunha**
As reported in #675, there is something super strange with EEGNetV4.

To begin with, we have a dead variable inside the Neural Network. The docstring isn't completed, not all the parameters are exposed, and a module that will be used by several models that are arriving is only in the eegnet folder.

In this PR I try to solve all these problems.

### CODE REVIEW

**Comment by bruAristimunha** [comment]
I want to rename the variable's names too:

F1: num_filters_temporal
D: depth_multiplier
F2: num_filters_pointwise

But it should be in another PR for this modification.

**Comment by bruAristimunha** [comment]
This issue is also reported here: #651; this PR should be a good place to fix/think this.

File: braindecode/models/modules.py [hunk 1]
```diff
@@ -854,3 +854,15 @@ def _apply_iir(x: Tensor, filter: dict) -> Tensor:
...
+
+class Conv2dWithConstraint(nn.Conv2d):
+    def __init__(self, *args, max_norm=1, **kwargs):
+        self.max_norm = max_norm
+        super(Conv2dWithConstraint, self).__init__(*args, **kwargs)
+
+    def forward(self, x):
+        self.weight.data = torch.renorm(
+            self.weight.data, p=2, dim=0, maxnorm=self.max_norm
+        )
```
Comment (tomMoral) [inline braindecode/models/modules.py:867]: Don't you need `torch.nograd` here?

File: braindecode/models/eegnet.py [hunk 2]
```diff
@@ -1,47 +1,63 @@
...
+    final_conv_length : int or "auto", default="auto"
+        Length of the final convolution layer. If "auto", it is set based on the n_times.
+    pool_mode : str, {"mean", "max"}, default="mean"
+        Pooling method to use in pooling layers.
+    F1 : int, default=8
+        Number of temporal filters in the first convolutional layer.
+    D : int, default=2
+        Depth multiplier for the depthwise convolution.
+    F2 : int, default=16
+        Number of pointwise filters in the separable convolution. Usually set to ``F1 * D``.
```
Comment (tomMoral) [inline braindecode/models/eegnet.py:35]: Do we keep this? or maybe use `default = None`, which default to `F1 * D`?

File: braindecode/models/eegnet.py [hunk 3]
```diff
@@ -1,47 +1,63 @@
...
+    pool1_stride_size : int, default=4
+        Stride size of the first pooling layer.
+    pool2_kernel_size : int, default=8
+        Kernel size of the second pooling layer.
+    pool2_stride_size : int, default=8
+        Stride size of the second pooling layer.
+    kernel_length : int, default=64
+        Length of the temporal convolution kernel.
+    drop_prob : float, default=0.25
+        Dropout probability.
```
Comment (tomMoral) [inline braindecode/models/eegnet.py:49]: Say where the dropout is applied.

File: braindecode/models/eegnet.py [hunk 4]
```diff
@@ -1,47 +1,63 @@
...
+    activation : nn.Module, default=nn.ELU
+        Activation function to apply. Should be a PyTorch activation module like
+        ``nn.ReLU`` or ``nn.ELU``.
```
Comment (tomMoral) [inline braindecode/models/eegnet.py:52]: Say where it is applied.

File: braindecode/models/eegnet.py [hunk 5]
```diff
@@ -1,47 +1,63 @@
...
+    batch_norm_momentum : float, default=0.01
+        Momentum for the batch normalization layers.
+    batch_norm_affine : bool, default=True
+        Whether to include learnable affine parameters in batch normalization layers.
+    batch_norm_eps : float, default=1e-3
+        Epsilon value for batch normalization layers.
+    conv_spatial_max_norm : float, default=1
+        Max norm constraint for the spatial convolution layer.
```
Comment (tomMoral) [inline braindecode/models/eegnet.py:60]: This should be above batch norm params.

File: braindecode/models/eegnet.py [hunk 6]
```diff
@@ -50,27 +66,35 @@ class EEGNetv4(EEGModuleMixin, nn.Sequential):
...
+        pool1_stride_size: int = 4,
+        pool2_kernel_size: int = 8,
+        pool2_stride_size: int = 8,
+        drop_prob: float = 0.25,
         activation: nn.Module = nn.ELU,
+        batch_norm_momentum: float = 0.01,
+        batch_norm_affine: bool = True,
+        batch_norm_eps: float = 1e-3,
+        conv_spatial_max_norm: int = 1,
         chs_info=None,
```
Comment (tomMoral) [inline braindecode/models/eegnet.py:98]: For deprecation, add a `**kwargs` argument so it can still be passed and raise a warning if it is not empty?

**Review by tomMoral — APPROVED** [review]
Few comments but overall LGTM.

**Review by tomMoral — APPROVED** [review]
LGTM!
